File: app-server-webflux/src/main/java/com/github/arhor/simple/expense/tracker/service/impl/UserServiceImpl.java
package com.github.arhor.simple.expense.tracker.service.impl;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import reactor.core.publisher.Mono;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
import org.springframework.stereotype.Service;

import com.github.arhor.simple.expense.tracker.data.model.InternalUser;
import com.github.arhor.simple.expense.tracker.data.repository.InternalUserRepository;
import com.github.arhor.simple.expense.tracker.exception.EntityDuplicateException;
import com.github.arhor.simple.expense.tracker.exception.EntityNotFoundException;
import com.github.arhor.simple.expense.tracker.model.Currency;
import com.github.arhor.simple.expense.tracker.model.UserRequestDTO;
import com.github.arhor.simple.expense.tracker.model.UserResponseDTO;
import com.github.arhor.simple.expense.tracker.service.UserService;
import com.github.arhor.simple.expense.tracker.service.mapper.InternalUserMapper;

@Slf4j
@Service
@RequiredArgsConstructor(onConstructor_ = @Autowired)
public class UserServiceImpl implements UserService {

    private final InternalUserRepository userRepository;
    private final InternalUserMapper userMapper;

    @Override
    public Mono<UserResponseDTO> createNewUser(final UserRequestDTO userDTO) {
        val username = userDTO.getUsername();

        return userRepository.existsByUsername(username).flatMap(exists -> {
            if (exists) {
                return Mono.error(
                    new EntityDuplicateException(
                        "InternalUser", "username=" + username
                    )
                );
            }

            val user = userMapper.mapToUser(userDTO);
            val createdUser = userRepository.save(user);

            return createdUser.map(userMapper::mapToResponse);
        });
    }

    @Override
    public Mono<Void> createNewUserIfNecessary(final Authentication authentication) {
        if (authentication instanceof OAuth2AuthenticationToken authenticationToken) {
            val extId = authenticationToken.getName();
            val extProvider = authenticationToken.getAuthorizedClientRegistrationId();

            if ((extId != null) && (extProvider != null)) {
                return userRepository.existsByExternalIdAndExternalProvider(extId, extProvider).flatMap(exists -> {
                    if (exists) {
                        return Mono.empty();
                    }
                    return userRepository.save(
                        InternalUser.builder()
                            .externalId(extId)
                            .externalProvider(extProvider)
                            .currency(Currency.USD.name())
                            .build()
                    ).then();
                });
            }
        }
        return Mono.empty();
    }

    @Override
    public Mono<UserResponseDTO> determineUser(final Authentication auth) {
        return determineInternalUser(auth).map(userMapper::mapToResponse);
    }

    private Mono<InternalUser.Projection> determineInternalUser(final Authentication auth) {
        if (auth instanceof OAuth2AuthenticationToken token) {
            val externalId = token.getName();
            val externalProvider = token.getAuthorizedClientRegistrationId();

            return userRepository.findByExternalIdAndProvider(externalId, externalProvider).switchIfEmpty(
                Mono.error(
                    () -> new EntityNotFoundException(
                        "User", "externalId=" + externalId + ", externalProvider=" + externalProvider
                    )
                )
            );
        }
        if (auth instanceof UsernamePasswordAuthenticationToken token) {
            val username = token.getName();

            return userRepository.findByUsername(username).switchIfEmpty(
                Mono.error(() ->
                    new EntityNotFoundException(
                        "User", "username=" + username
                    )
                )
            );
        }
        return Mono.error(() ->
            new IllegalArgumentException(
                "Unsupported authentication type: " + (
                    (auth != null)
                        ? auth.getClass().getSimpleName()
                        : "<null>"
                )
            )
        );
    }
}

File: apps/server/src/main/kotlin/com/github/arhor/simple/expense/tracker/service/impl/UserServiceImpl.kt
package com.github.arhor.simple.expense.tracker.service.impl

import com.github.arhor.simple.expense.tracker.data.model.InternalUser
import com.github.arhor.simple.expense.tracker.data.repository.InternalUserRepository
import com.github.arhor.simple.expense.tracker.service.exception.EntityDuplicateException
import com.github.arhor.simple.expense.tracker.service.exception.EntityNotFoundException
import com.github.arhor.simple.expense.tracker.model.Currency
import com.github.arhor.simple.expense.tracker.model.UserRequestDTO
import com.github.arhor.simple.expense.tracker.model.UserResponseDTO
import com.github.arhor.simple.expense.tracker.service.CustomUserDetails
import com.github.arhor.simple.expense.tracker.service.UserService
import com.github.arhor.simple.expense.tracker.service.mapping.InternalUserMapper
import org.springframework.data.repository.findByIdOrNull
import org.springframework.security.core.Authentication
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class UserServiceImpl(
    private val passwordEncoder: PasswordEncoder,
    private val userRepository: InternalUserRepository,
    private val userMapper: InternalUserMapper,
) : UserService {

    override fun loadUserByUsername(username: String): CustomUserDetails {
        return userRepository.findInternalUserByUsername(username)?.let(userMapper::mapToUserDetails)
            ?: throw UsernameNotFoundException("InternalUser with username '$username' is not found.")
    }

    override fun getUserById(userId: Long): UserResponseDTO {
        return userRepository.findByIdOrNull(userId)?.let(userMapper::mapToUserResponse)
            ?: throw EntityNotFoundException(InternalUser.ENTITY_NAME, "id=${userId}")
    }

    @Transactional
    override fun createNewUser(request: UserRequestDTO): UserResponseDTO {
        val username = request.username

        if (userRepository.existsByUsername(username)) {
            throw EntityDuplicateException(InternalUser.ENTITY_NAME, "username=$username")
        }

        request.apply {
            password = passwordEncoder.encode(password)
        }

        val user = userMapper.mapToInternalUser(request)
        val createdUser = userRepository.save(user)

        return userMapper.mapToUserResponse(createdUser)
    }

    @Override
    override fun createInternalUserIfNecessary(auth: Authentication) {
        if (auth !is OAuth2AuthenticationToken) {
            return
        }

        val externalId = auth.name
        val externalProvider = auth.authorizedClientRegistrationId

        val internalUserAlreadyExists =
            userRepository.existsByExternalIdAndExternalProvider(
                externalId,
                externalProvider,
            )

        if (!internalUserAlreadyExists) {
            userRepository.save(
                InternalUser(
                    externalId = externalId,
                    externalProvider = externalProvider,
                    currency = Currency.USD.name,
                )
            )
        }
    }
}
